### PR TITLE
feat!: Support only sending a subset of the fields to OPA

### PR DESCRIPTION
**Pull request by sbernauer**
# Description

One step of working on https://github.com/stackabletech/hdfs-utils/issues/48

There is a branch to build a custom image: https://github.com/stackabletech/docker-images/tree/feat/hdfs-opa-performance

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes

```[tasklist]
# Reviewer
- [x] Code contains useful comments
- [ ] (Integration-)Test cases added
- [ ] Documentation added or updated
- [x] Changelog updated
```

```[tasklist]
# Acceptance
- [ ] Proper release label has been added
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -26,6 +26,8 @@ The Stackable HDFS already takes care of this, you don't need to do anything in
 
 - Set `dfs.namenode.inode.attributes.provider.class` in `hdfs-site.xml` to `tech.stackable.hadoop.StackableAuthorizer`
 - Set `hadoop.security.authorization.opa.policy.url` in `core-site.xml` to the HTTP endpoint of your OPA rego rule, e.g. `http://opa.default.svc.cluster.local:8081/v1/data/hdfs/allow`
+- The property `hadoop.security.authorization.opa.extended-requests` (defaults to `false`), which controls if all fields should be send to OPA.
```
Comment (lfrancke) [inline README.md]:
```suggestion
- The property `hadoop.security.authorization.opa.extended-requests` (defaults to `false`), which controls if all fields should be sent to OPA.
```

Comment (lfrancke) [reply README.md]: This needs documentation somewhere which fields it will always send.

File: pom.xml [hunk 2]
```diff
@@ -49,8 +50,8 @@
     <maven-site-plugin.version>3.12.1</maven-site-plugin.version>
     <maven-surefire-plugin.version>3.3.1</maven-surefire-plugin.version>
     <spotless-maven-plugin.version>2.43.0</spotless-maven-plugin.version>
-    <kubernetes-client.version>6.13.1</kubernetes-client.version>
-    <okio.version>1.17.6</okio.version>
+    <kubernetes-client.version>6.13.4</kubernetes-client.version>
+    <okio.version>3.9.1</okio.version>
```
Comment (lfrancke) [inline pom.xml]: This is not necessarily compatible. I'd recommend keeping it at 1.17.6 for now.

File: src/main/java/tech/stackable/hadoop/StackableAccessControlEnforcer.java [hunk 3]
```diff
@@ -60,7 +62,14 @@ public StackableAccessControlEnforcer() {
         try {
             this.opaUri = URI.create(opaPolicyUrl);
         } catch (Exception e) {
-            throw new OpaException.UriInvalid(opaUri, e);
+            throw new OpaException.UriInvalid(opaPolicyUrl, e);
+        }
+
+        String extendedRequests = configuration.get(EXTENDED_REQUESTS_PROP, "false");
```
Comment (lfrancke) [inline src/main/java/tech/stackable/hadoop/StackableAccessControlEnforcer.java]: There is a `configuration.getBoolean`

File: src/main/java/tech/stackable/hadoop/StackableAccessControlEnforcer.java [hunk 4]
```diff
@@ -78,7 +87,14 @@ public StackableAccessControlEnforcer() {
                 // Otherwise we get com.fasterxml.jackson.databind.JsonMappingException: Infinite recursion (StackOverflowError)
                 .addMixIn(DatanodeDescriptor.class, DatanodeDescriptorMixin.class);
 
-        LOG.debug("Started HdfsOpaAccessControlEnforcer");
+        StringBuilder logStartupStatement = new StringBuilder("Started HdfsOpaAccessControlEnforcer");
```
Comment (lfrancke) [inline src/main/java/tech/stackable/hadoop/StackableAccessControlEnforcer.java]:
```suggestion
        StringBuilder logStartupStatement = new StringBuilder("Started StackableAccessControlEnforcer");
```

**Review by lfrancke — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 2 above)
Comment (sbernauer) [reply pom.xml]: Removed all dependecy bumps from this PR, as we had #51 for that

File: src/main/java/tech/stackable/hadoop/StackableAccessControlEnforcer.java (hunk 3 above)
Comment (sbernauer) [reply src/main/java/tech/stackable/hadoop/StackableAccessControlEnforcer.java]: Updated the PR to use that, thanks!

File: README.md (hunk 1 above)
Comment (sbernauer) [reply README.md]: Please have a look at e2124091afadeab7a22f7f5ab155e1628fce0779

File: src/main/java/tech/stackable/hadoop/StackableAccessControlEnforcer.java [hunk 5]
```diff
@@ -197,7 +170,21 @@ public void checkPermission(
...
       throws AccessControlException {
-    OpaAllowQuery query = new OpaAllowQuery(new OpaAllowQuery.OpaAllowQueryInput(authzContext));
+    // When executing "hdfs dfs -du /" the path is set to null. This does not worsen security, as
+    // "/" is the highest level of access that a user can have.
+    if (authzContext.getOperationName().equals("contentSummary")
+        && authzContext.getPath() == null) {
+      authzContext.setPath("/");
+    }
+
+    final Object query;
```
Comment (lfrancke) [inline src/main/java/tech/stackable/hadoop/StackableAccessControlEnforcer.java]:
```suggestion
    Object query;
```

We don't use final elsewhere

File: README.md [hunk 6]
```diff
@@ -26,13 +26,51 @@ The Stackable HDFS already takes care of this, you don't need to do anything in
 
 - Set `dfs.namenode.inode.attributes.provider.class` in `hdfs-site.xml` to `tech.stackable.hadoop.StackableAuthorizer`
 - Set `hadoop.security.authorization.opa.policy.url` in `core-site.xml` to the HTTP endpoint of your OPA rego rule, e.g. `http://opa.default.svc.cluster.local:8081/v1/data/hdfs/allow`
+- The property `hadoop.security.authorization.opa.extended-requests` (defaults to `false`), which controls if all fields should be sent to OPA.
```
Comment (lfrancke) [inline README.md]:
```suggestion
- The property `hadoop.security.authorization.opa.extended-requests` (defaults to `false`) controls if all fields (`true`) should be sent to OPA or only a subset
```

**Review by lfrancke — CHANGES_REQUESTED** [review]
Only two minor things left

**Review by lfrancke — APPROVED** [review]

**Pull request merged**